packages/dates/src/format.ts

```diff
@@ -0,0 +1,201 @@
+import {memoizedGetDateTimeFormat} from './utilities';
+
+export function format(
+  date: Date,
+  fmtStr: string,
+  tz?: string,
+  localeStr?: string,
+): string {
+  const options = memoizedGetDateTimeFormat(localeStr, {
+    timeZone: tz,
+  }).resolvedOptions();
+  let result = fmtStr;
+  result = replaceToken(result, 'YYYY', getYear(date, 'numeric', options));
```

Comment:
This looks like you're calling the get* function even if the token was not found in the fmtStr that was requested. This means a `format(SOMEDATE, '')` shall make 20 calls to IntlDateTime. toLocaleTimeString / toLocaleDateString (as each get* call makes one request). That seems absurdly inefficient. You'll need to make this smart enough request the minimal amount of toLocaleTimeString / toLocaleDateString requests.

Reply:
Thanks for the suggestion. I modified the function to use `Intl.DateTimeFormat().formatToParts` like you suggested and now it also only calls the aforementioned function for each given token in the format string (as opposed to calling it 20 times for every string).

I had to call it for each token as opposed to calling it once with all the parts in the format string and pulling the parts out because some of the parts get modified by the presence of the others. Thus to get accurate reads for each part, I have to get each part in isolation from the others.